File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.py
def file_filter(file):
    HEAD = ""
    HEAD_flag = False
    NUMBER = ""
    NUMBER_flag = False
    TAIL = ""
    number_index = 0
    tail_index = 0
    
    for i in range(len(file)):
        if not HEAD_flag:
            if file[i].isdigit():
                HEAD_flag = True
            else:
                HEAD += file[i]
        
        if HEAD_flag and not NUMBER_flag:
            if file[i].isdigit():
                NUMBER += file[i]
            else:
                NUMBER_flag = True
        
        if HEAD_flag and NUMBER_flag:
            TAIL += file[i]
        
    return [HEAD, NUMBER, TAIL]

def solution(files):
    answer = []
    result = []
    for file in files:
        result.append(file_filter(file))
    
    result.sort(key = lambda x : [x[0].upper(), int(x[1])])
    
    for file in result:
        answer.append("".join(file))
    
    return answer

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.js
function solution(files) {
    const convertedFiles = [];

    for (const file of files) {
        let HEAD = "";
        let NUMBER = "";
        let TAIL = "";
        let numberStarted = false; // NUMBER 구간 시작 여부
        let numberEnded = false;   // NUMBER 구간 종료 여부 (TAIL 진입)

        for (const s of file) {
            const isDigit = !isNaN(s) && s !== " ";

            if (!numberEnded && isDigit) {
                // 아직 NUMBER 구간이 끝나지 않았고, 숫자라면 NUMBER에 추가
                numberStarted = true;
                NUMBER += s;
            } else if (!numberStarted) {
                // 아직 숫자를 한 번도 못 만났으면 HEAD
                HEAD += s;
            } else {
                // NUMBER를 이미 지나온 상태라면(숫자든 아니든) TAIL
                numberEnded = true;
                TAIL += s;
            }
        }

        convertedFiles.push([HEAD, NUMBER, TAIL]);
    }

    const answer = convertedFiles
        .map((file, idx) => [...file, idx]) // 원래 순서 보존용 인덱스 추가
        .sort((a, b) => {
            const headerA = a[0].toUpperCase();
            const headerB = b[0].toUpperCase();
            if (headerA < headerB) return -1;
            if (headerA > headerB) return 1;

            const numberA = Number(a[1]);
            const numberB = Number(b[1]);
            if (numberA !== numberB) return numberA - numberB;

            return a[3] - b[3]; // 안정 정렬 보장
        })
        .map((file) => file[0] + file[1] + file[2]);

    return answer;
}